apollo-router/src/plugins/telemetry/metrics.rs

```diff
@@ -0,0 +1,311 @@
+use apollo_router_core::{
+    http_compat, Handler, Plugin, ResponseBody, RouterRequest, RouterResponse, SubgraphRequest,
+    SubgraphResponse,
+};
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use http::{Method, StatusCode};
+use opentelemetry::{
+    global,
+    metrics::{Counter, ValueRecorder},
+    sdk::metrics::PushController,
+    KeyValue,
+};
+use opentelemetry_prometheus::PrometheusExporter;
+use prometheus::{Encoder, Registry, TextEncoder};
+use reqwest::Url;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_json_bytes::from_value;
+use std::{
+    task::{Context, Poll},
+    time::SystemTime,
+};
+use tower::{service_fn, steer::Steer, util::BoxService, BoxError, Service, ServiceExt};
+
+#[cfg(any(feature = "otlp-grpc", feature = "otlp-http"))]
+use super::otlp::Metrics as OltpConfiguration;
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+pub struct MetricsConfiguration {
+    pub exporter: MetricsExporter,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+#[serde(deny_unknown_fields, rename_all = "snake_case")]
+pub enum MetricsExporter {
+    Prometheus(PrometheusConfiguration),
+    #[cfg(any(feature = "otlp-grpc", feature = "otlp-http"))]
+    Otlp(Box<OltpConfiguration>),
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, JsonSchema)]
+pub struct PrometheusConfiguration {
+    endpoint: String,
+}
+
+#[derive(Debug)]
+pub struct MetricsPlugin {
+    exporter: Option<PrometheusExporter>,
+    conf: MetricsConfiguration,
+    metrics_controller: Option<PushController>,
+    router_metrics: BasicMetrics,
+    subgraph_metrics: BasicMetrics,
+}
+
+#[derive(Debug)]
+pub struct BasicMetrics {
+    http_requests_total: Counter<u64>,
+    http_requests_error_total: Counter<u64>,
+    http_requests_duration: ValueRecorder<f64>,
+}
+
+#[async_trait::async_trait]
+impl Plugin for MetricsPlugin {
+    type Config = MetricsConfiguration;
+
+    fn new(config: Self::Config) -> Result<Self, BoxError> {
+        let exporter = opentelemetry_prometheus::exporter().init();
+        let meter = global::meter("apollo/router");
+
+        if let MetricsExporter::Prometheus(prom_exporter_cfg) = &config.exporter {
+            if Url::parse(&format!("http://test:8080{}", prom_exporter_cfg.endpoint)).is_err() {
```

Comment:
why is there a test url here?

Reply:
We don't really care about this url, it's just a fake URL to test that the endpoint passed in the config is compliant to be used as a path in an URL